File: Android/app/src/main/java/com/kawai/mochi/StickerPackListActivity.java
package com.kawai.mochi;

import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.util.Pair;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;


public class StickerPackListActivity extends AddStickerPackActivity {
    public static final String EXTRA_STICKER_PACK_LIST_DATA = "sticker_pack_list";
    private static final int STICKER_PREVIEW_DISPLAY_LIMIT = 5;
    private static final ExecutorService executor = Executors.newCachedThreadPool();
    private static final Handler mainHandler = new Handler(Looper.getMainLooper());
    private LinearLayoutManager packLayoutManager;
    private RecyclerView packRecyclerView;
    private StickerPackListAdapter allStickerPacksListAdapter;
    private volatile boolean whitelistCheckCancelled;
    private ArrayList<StickerPack> stickerPackList;
    private View emptyStateLayout;
    private ExtendedFloatingActionButton importFab;
    private ActivityResultLauncher<Intent> filePickerLauncher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sticker_pack_list);

        androidx.appcompat.widget.Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        packRecyclerView = findViewById(R.id.sticker_pack_list);
        emptyStateLayout = findViewById(R.id.empty_state_layout);
        importFab = findViewById(R.id.import_button_fab);

        stickerPackList = getIntent().getParcelableArrayListExtra(EXTRA_STICKER_PACK_LIST_DATA);
        if (stickerPackList == null) stickerPackList = new ArrayList<>();

        if (getSupportActionBar() != null) {
            getSupportActionBar().setTitle(getResources().getQuantityString(R.plurals.title_activity_sticker_packs_list, stickerPackList.size()));
        }

        // Setup file picker for .wasticker import
        filePickerLauncher = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                result -> {
                    if (result.getResultCode() == RESULT_OK && result.getData() != null) {
                        Uri uri = result.getData().getData();
                        if (uri != null) {
                            importWastickerFile(uri);
                        }
                    }
                });

        showStickerPackList(stickerPackList);

        // Empty state import button
        Button emptyImportButton = findViewById(R.id.import_button);
        if (emptyImportButton != null) {
            emptyImportButton.setOnClickListener(v -> openFilePicker());
        }

        // FAB import button
        importFab.setOnClickListener(v -> openFilePicker());

        // Swipe to delete
        new ItemTouchHelper(new ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT | ItemTouchHelper.RIGHT) {
            @Override
            public boolean onMove(@NonNull RecyclerView recyclerView, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull RecyclerView.ViewHolder target) {
                return false;
            }

            @Override
            public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
                int position = viewHolder.getAdapterPosition();
                if (position >= 0 && position < stickerPackList.size()) {
                    StickerPack pack = stickerPackList.get(position);
                    new MaterialAlertDialogBuilder(StickerPackListActivity.this)
                            .setTitle("Delete Pack")
                            .setMessage("Delete \"" + pack.name + "\"?")
                            .setPositiveButton("Delete", (dialog, which) -> deletePack(position))
                            .setNegativeButton("Cancel", (dialog, which) -> {
                                allStickerPacksListAdapter.notifyItemChanged(position);
                            })
                            .setOnCancelListener(d -> allStickerPacksListAdapter.notifyItemChanged(position))
                            .show();
                }
            }
        }).attachToRecyclerView(packRecyclerView);

        updateEmptyState();
    }

    private void openFilePicker() {
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("*/*");
        intent.addCategory(Intent.CATEGORY_OPENABLE);
        filePickerLauncher.launch(Intent.createChooser(intent, "Select .wasticker file"));
    }

    private void importWastickerFile(Uri uri) {
        WeakReference<StickerPackListActivity> ref = new WeakReference<>(this);
        executor.execute(() -> {
            StickerPackListActivity activity = ref.get();
            if (activity == null) return;
            String error = null;
            ArrayList<StickerPack> freshPacks = null;
            try {
                WastickerParser.importStickerPack(activity, uri);
                StickerContentProvider provider = StickerContentProvider.getInstance();
                if (provider != null) provider.invalidateStickerPackList();
                // Fetch fresh list on the background thread — zero main-thread work
                freshPacks = StickerPackLoader.fetchStickerPacks(activity);
            } catch (Exception e) {
                error = e.getMessage();
            }
            final String finalError = error;
            final ArrayList<StickerPack> finalPacks = freshPacks;
            mainHandler.post(() -> {
                StickerPackListActivity act = ref.get();
                if (act == null) return;
                if (finalError != null) {
                    Toast.makeText(act, "Import error: " + finalError, Toast.LENGTH_LONG).show();
                } else {
                    // Update list in-place — user never sees a reload
                    Toast.makeText(act, "✓ Pack imported!", Toast.LENGTH_SHORT).show();
                    if (finalPacks != null) {
                        act.stickerPackList.clear();
                        act.stickerPackList.addAll(finalPacks);
                        // Ensure adapter is using the updated list object
                        act.allStickerPacksListAdapter.setStickerPackList(act.stickerPackList);
                        act.allStickerPacksListAdapter.notifyDataSetChanged();
                        act.updateEmptyState();
                        if (act.getSupportActionBar() != null) {
                            act.getSupportActionBar().setTitle(act.getResources()
                                .getQuantityString(R.plurals.title_activity_sticker_packs_list,
                                    act.stickerPackList.size()));
                        }
                        // Scroll to show the newly imported pack
                        act.packRecyclerView.smoothScrollToPosition(act.stickerPackList.size() - 1);
                    }
                }
            });
        });
    }

    private void deletePack(int position) {
        if (position < 0 || position >= stickerPackList.size()) return;
        StickerPack pack = stickerPackList.get(position);
        WeakReference<StickerPackListActivity> ref = new WeakReference<>(this);
        executor.execute(() -> {
            StickerPackListActivity activity = ref.get();
            if (activity == null) return;
            String error;
            try {
                WastickerParser.deleteStickerPack(activity, pack.identifier);
                StickerContentProvider provider = StickerContentProvider.getInstance();
                if (provider != null) provider.invalidateStickerPackList();
                error = null;
            } catch (Exception e) {
                error = e.getMessage();
            }
            final String finalError = error;
            mainHandler.post(() -> {
                StickerPackListActivity act = ref.get();
                if (act == null) return;
                if (finalError != null) {
                    Toast.makeText(act, "Error: " + finalError, Toast.LENGTH_LONG).show();
                    act.allStickerPacksListAdapter.notifyItemChanged(position);
                } else {
                    // Critical: remove from the actual list object the adapter is using
                    act.stickerPackList.remove(position);
                    act.allStickerPacksListAdapter.notifyItemRemoved(position);
                    // Update remaining item indices to prevent ghosting or index mismatch
                    if (position < act.stickerPackList.size()) {
                        act.allStickerPacksListAdapter.notifyItemRangeChanged(position, act.stickerPackList.size() - position);
                    }
                    act.updateEmptyState();
                    if (act.getSupportActionBar() != null) {
                        act.getSupportActionBar().setTitle(act.getResources()
                                .getQuantityString(R.plurals.title_activity_sticker_packs_list,
                                        act.stickerPackList.size()));
                    }
                    Toast.makeText(act, "Pack deleted", Toast.LENGTH_SHORT).show();
                }
            });
        });
    }

    private void updateEmptyState() {
        if (stickerPackList.isEmpty()) {
            packRecyclerView.setVisibility(View.GONE);
            emptyStateLayout.setVisibility(View.VISIBLE);
            importFab.setVisibility(View.GONE);
        } else {
            packRecyclerView.setVisibility(View.VISIBLE);
            emptyStateLayout.setVisibility(View.GONE);
            importFab.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        menu.add(Menu.NONE, android.R.id.home, Menu.NONE, "Settings")
                .setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if ("Settings".equals(item.getTitle())) {
            startActivity(new Intent(this, SettingsActivity.class));
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onResume() {
        super.onResume();
        whitelistCheckCancelled = false;
        WeakReference<StickerPackListActivity> ref = new WeakReference<>(this);
        executor.execute(() -> {
            if (whitelistCheckCancelled) return;
            StickerPackListActivity activity = ref.get();
            if (activity == null) return;
            // Directly update the objects in the existing list to maintain reference integrity
            for (StickerPack stickerPack : activity.stickerPackList) {
                stickerPack.setIsWhitelisted(WhitelistCheck.isWhitelisted(activity, stickerPack.identifier));
            }
            mainHandler.post(() -> {
                if (whitelistCheckCancelled) return;
                StickerPackListActivity act = ref.get();
                if (act != null) {
                    act.allStickerPacksListAdapter.notifyDataSetChanged();
                }
            });
        });
    }

    @Override
    protected void onPause() {
        super.onPause();
        whitelistCheckCancelled = true;
    }

    private android.view.ViewTreeObserver.OnGlobalLayoutListener globalLayoutListener;

    private void showStickerPackList(List<StickerPack> stickerPackList) {
        allStickerPacksListAdapter = new StickerPackListAdapter(stickerPackList, onAddButtonClickedListener);
        packRecyclerView.setAdapter(allStickerPacksListAdapter);
        packLayoutManager = new LinearLayoutManager(this);
        packLayoutManager.setOrientation(RecyclerView.VERTICAL);
        packRecyclerView.setLayoutManager(packLayoutManager);
        // Increase off-screen view cache to avoid rebinds while flinging
        packRecyclerView.setItemViewCacheSize(12);
        globalLayoutListener = this::recalculateColumnCount;
        packRecyclerView.getViewTreeObserver().addOnGlobalLayoutListener(globalLayoutListener);
    }

    private final StickerPackListAdapter.OnAddButtonClickedListener onAddButtonClickedListener = pack -> addStickerPackToWhatsApp(pack.identifier, pack.name);

    private void recalculateColumnCount() {
        final int rvWidth = packRecyclerView.getWidth();
        if (rvWidth <= 0) return; // RecyclerView not measured yet — wait for next pass

        // imageRowView width = rvWidth - card margins (12dp×2) - ConstraintLayout padding (16dp×2) - add button (48dp)
        // This avoids the chicken-and-egg where imageRowView.width==0 because it is GONE until maxColumns is set.
        final float density = getResources().getDisplayMetrics().density;
        final int overhead = Math.round((12 + 12 + 16 + 16 + 48) * density);
        final int imageRowWidth = rvWidth - overhead;
        if (imageRowWidth <= 0) return;

        final int previewSize = getResources().getDimensionPixelSize(R.dimen.sticker_pack_list_item_preview_image_size);
        int maxNumberOfImagesInARow = Math.min(STICKER_PREVIEW_DISPLAY_LIMIT, Math.max(imageRowWidth / previewSize, 1));
        int minMarginBetweenImages = maxNumberOfImagesInARow > 1
                ? (imageRowWidth - maxNumberOfImagesInARow * previewSize) / (maxNumberOfImagesInARow - 1)
                : 0;
        allStickerPacksListAdapter.setImageRowSpec(maxNumberOfImagesInARow, minMarginBetweenImages);
        // Remove after first valid measurement — layout width won't change
        if (packRecyclerView.getViewTreeObserver().isAlive()) {
            packRecyclerView.getViewTreeObserver().removeOnGlobalLayoutListener(globalLayoutListener);
        }
    }

}

File: Android/app/src/main/java/com/kawai/mochi/StickerPackListActivity.kt
package com.kawai.mochi

import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.widget.Toolbar
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.repeatOnLifecycle
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.floatingactionbutton.ExtendedFloatingActionButton
import com.google.android.material.progressindicator.LinearProgressIndicator
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlin.Pair
import kotlin.math.roundToInt


class StickerPackListActivity : AddStickerPackActivity(), ThumbnailRegenerationManager.Listener {
    private lateinit var packLayoutManager: LinearLayoutManager
    private lateinit var packRecyclerView: RecyclerView
    private lateinit var allStickerPacksListAdapter: StickerPackListAdapter
    private var stickerPackList = ArrayList<StickerPack>()
    private lateinit var emptyStateLayout: View
    private lateinit var importFab: ExtendedFloatingActionButton
    private lateinit var filePickerLauncher: ActivityResultLauncher<Intent>
    private lateinit var swipeRefresh: SwipeRefreshLayout
    private lateinit var itemTouchHelper: ItemTouchHelper
    
    private var importProgressContainer: View? = null
    private var importProgressBar: LinearProgressIndicator? = null
    private var importStatusText: TextView? = null

    private val telegramImportLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            refreshStickerPacks()
        }
    }

    private val scrollListener = object : RecyclerView.OnScrollListener() {
        override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
            // Feature removed: no longer notifying adapter of scroll state
        }

        override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
            super.onScrolled(recyclerView, dx, dy)
            if (dy > 0 && importFab.isExtended) importFab.shrink() 
            else if (dy < 0 && !importFab.isExtended) importFab.extend()
        }
    }

    private val settingsLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { 
        allStickerPacksListAdapter.invalidateAnimationsCache()
        refreshStickerPacks()
    }

    private val mergePacksLauncher = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result ->
        if (result.resultCode == RESULT_OK) {
            refreshStickerPacks()
        }
    }

    override fun showProgressBar(message: String?) {
        importProgressContainer?.visibility = View.VISIBLE
        importProgressBar?.isIndeterminate = true
        importStatusText?.text = message ?: getString(R.string.add_to_whatsapp)
    }

    override fun hideProgressBar() {
        if (!ThumbnailRegenerationManager.isRegenerating()) {
            importProgressContainer?.visibility = View.GONE
        }
    }

    override fun updateProgress(current: Int, total: Int, message: String?) {
        importProgressContainer?.visibility = View.VISIBLE
        importProgressBar?.isIndeterminate = false
        importProgressBar?.max = total
        importProgressBar?.setProgressCompat(current, true)
        if (message != null) {
            importStatusText?.text = message
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sticker_pack_list)

        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)

        swipeRefresh = findViewById(R.id.swipe_refresh)
        packRecyclerView = findViewById(R.id.sticker_pack_list)
        emptyStateLayout = findViewById(R.id.empty_state_layout)
        importFab = findViewById(R.id.import_button_fab)
        
        importProgressContainer = findViewById(R.id.import_progress_container)
        importProgressBar = findViewById(R.id.import_progress_bar)
        importStatusText = findViewById(R.id.import_status_text)

        val intentList = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            intent.getParcelableArrayListExtra(EXTRA_STICKER_PACK_LIST_DATA, StickerPack::class.java)
        } else {
            @Suppress("DEPRECATION")
            intent.getParcelableArrayListExtra(EXTRA_STICKER_PACK_LIST_DATA)
        }
        
        stickerPackList = intentList ?: ArrayList()

        supportActionBar?.title = resources.getQuantityString(R.plurals.title_activity_sticker_packs_list, stickerPackList.size)

        filePickerLauncher = registerForActivityResult(ActivityResultContracts.StartActivityForResult()) { result ->
            if (result.resultCode == RESULT_OK) {
                result.data?.data?.let { uri ->
                    importWastickerFile(uri)
                }
            }
        }

        showStickerPackList(stickerPackList)

        findViewById<Button>(R.id.import_button)?.setOnClickListener { showImportChoice() }
        importFab.setOnClickListener { showImportChoice() }

        swipeRefresh.setOnRefreshListener {
            refreshStickerPacks()
        }

        itemTouchHelper = ItemTouchHelper(object : ItemTouchHelper.SimpleCallback(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT) {
            override fun onMove(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder, target: RecyclerView.ViewHolder) = false
            override fun onSelectedChanged(viewHolder: RecyclerView.ViewHolder?, actionState: Int) {
                super.onSelectedChanged(viewHolder, actionState)
                swipeRefresh.isEnabled = actionState == ItemTouchHelper.ACTION_STATE_IDLE
            }
            override fun clearView(recyclerView: RecyclerView, viewHolder: RecyclerView.ViewHolder) {
                super.clearView(recyclerView, viewHolder)
                swipeRefresh.isEnabled = true
            }
            override fun onSwiped(viewHolder: RecyclerView.ViewHolder, direction: Int) {
                val position = viewHolder.bindingAdapterPosition
                if (position in stickerPackList.indices) {
                    val pack = stickerPackList[position]
                    MaterialAlertDialogBuilder(this@StickerPackListActivity)
                        .setTitle(R.string.delete_pack_title)
                        .setMessage(getString(R.string.delete_pack_confirm_with_name, pack.name))
                        .setPositiveButton(R.string.delete_button) { _, _ -> deletePack(position) }
                        .setNegativeButton(R.string.cancel) { _, _ -> cancelSwipeDelete(position) }
                        .setOnCancelListener { cancelSwipeDelete(position) }
                        .show()
                }
            }
        })
        itemTouchHelper.attachToRecyclerView(packRecyclerView)

        updateEmptyState()
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                StickerUpdateManager.updateEvent.collect {
                    refreshStickerPacks()
                }
            }
        }
        
        // Listen for thumbnail regeneration progress
        ThumbnailRegenerationManager.addListener(this)
    }

    override fun onDestroy() {
        ThumbnailRegenerationManager.removeListener(this)
        super.onDestroy()
    }

    // ThumbnailRegenerationManager.Listener implementation
    override fun onProgress(current: Int, total: Int) {
        runOnUiThread {
            importProgressContainer?.visibility = View.VISIBLE
            importProgressBar?.isIndeterminate = false
            importProgressBar?.max = total
            importProgressBar?.setProgressCompat(current, true)
            importStatusText?.text = getString(R.string.progress_format, current, total)
        }
    }

    override fun onFinished() {
        runOnUiThread {
            importProgressContainer?.visibility = View.GONE
            refreshStickerPacks()
        }
    }

    override fun onError(message: String) {
        runOnUiThread {
            importProgressContainer?.visibility = View.GONE
            Toast.makeText(this, message, Toast.LENGTH_LONG).show()
        }
    }

    private fun refreshStickerPacks() {
        lifecycleScope.launch {
            try {
                val freshPacks = withContext(Dispatchers.IO) {
                    StickerPackLoader.fetchStickerPacks(this@StickerPackListActivity)
                }
                stickerPackList.clear()
                stickerPackList.addAll(freshPacks)

                if (::allStickerPacksListAdapter.isInitialized) {
                    allStickerPacksListAdapter.submitList(ArrayList(stickerPackList))
                }
                
                updateEmptyState()
                supportActionBar?.title = resources.getQuantityString(R.plurals.title_activity_sticker_packs_list, stickerPackList.size)
                
            } catch (e: Exception) {
                Log.e("ListActivity", "Refresh failed", e)
            } finally {
                swipeRefresh.isRefreshing = false
            }
        }
    }

    private fun cancelSwipeDelete(position: Int) {
        val viewHolder = packRecyclerView.findViewHolderForAdapterPosition(position)
        if (viewHolder != null) {
            ItemTouchHelper.Callback.getDefaultUIUtil().clearView(viewHolder.itemView)
        }
        allStickerPacksListAdapter.notifyItemChanged(position)
    }


    private fun showImportChoice() {
        val sheet = ImportChoiceBottomSheet.newInstance()
        sheet.setListener(object : ImportChoiceBottomSheet.Listener {
            override fun onImportFromFile() { openFilePicker() }
            override fun onImportFromTelegram() {
                telegramImportLauncher.launch(
                    Intent(this@StickerPackListActivity, TelegramImportActivity::class.java)
                )
            }
        })
        sheet.show(supportFragmentManager, "import_choice")
    }

    private fun openFilePicker() {
        val intent = Intent(Intent.ACTION_GET_CONTENT).apply {
            type = "*/*"
            addCategory(Intent.CATEGORY_OPENABLE)
        }
        filePickerLauncher.launch(Intent.createChooser(intent, getString(R.string.select_wasticker_file)))
    }

    private fun importWastickerFile(uri: Uri) {
        importProgressContainer?.visibility = View.VISIBLE
        importProgressBar?.isIndeterminate = true
        importStatusText?.setText(R.string.importing_pack)
        
        lifecycleScope.launch {
            try {
                val imported = withContext(Dispatchers.IO) {
                    val importedId = WastickerParser.importStickerPack(this@StickerPackListActivity, uri) { current, total ->
                        runOnUiThread {
                            importProgressBar?.isIndeterminate = false
                            importProgressBar?.max = total
                            importProgressBar?.setProgressCompat(current, true)
                            importStatusText?.text = getString(R.string.import_progress, current, total)
                        }
                    }
                    StickerContentProvider.getInstance()?.invalidateStickerPackList()
                    val importedPack = if (!importedId.isNullOrBlank()) {
                        StickerPackLoader.fetchStickerPack(this@StickerPackListActivity, importedId)
                    } else {
                        null
                    }
                    Pair(importedId, importedPack)
                }

                Toast.makeText(this@StickerPackListActivity, R.string.pack_imported, Toast.LENGTH_SHORT).show()

                val importedId = imported.first
                val importedPack = imported.second
                if (importedPack != null) {
                    val existingIndex = stickerPackList.indexOfFirst { it.identifier == importedId }
                    if (existingIndex >= 0) {
                        stickerPackList[existingIndex] = importedPack
                    } else {
                        stickerPackList.add(importedPack)
                    }
                    allStickerPacksListAdapter.submitList(ArrayList(stickerPackList))
                } else {
                    refreshStickerPacks()
                }

                updateEmptyState()
                supportActionBar?.title = resources.getQuantityString(R.plurals.title_activity_sticker_packs_list, stickerPackList.size)
                packRecyclerView.smoothScrollToPosition(stickerPackList.size - 1)
            } catch (e: Exception) {
                Toast.makeText(this@StickerPackListActivity, getString(R.string.import_error, e.message), Toast.LENGTH_LONG).show()
            } finally {
                if (!ThumbnailRegenerationManager.isRegenerating()) {
                    importProgressContainer?.visibility = View.GONE
                }
            }
        }
    }

    private fun deletePack(position: Int) {
        if (position !in stickerPackList.indices) return
        val pack = stickerPackList[position]

        stickerPackList.removeAt(position)
        allStickerPacksListAdapter.submitList(ArrayList(stickerPackList))
        updateEmptyState()
        supportActionBar?.title = resources.getQuantityString(
            R.plurals.title_activity_sticker_packs_list, stickerPackList.size)

        lifecycleScope.launch {
            try {
                withContext(Dispatchers.IO) {
                    WastickerParser.deleteStickerPack(this@StickerPackListActivity, pack.identifier)
                    StickerContentProvider.getInstance()?.invalidateStickerPackList()
                }
                Toast.makeText(this@StickerPackListActivity, R.string.pack_deleted, Toast.LENGTH_SHORT).show()
            } catch (e: Exception) {
                stickerPackList.add(position.coerceAtMost(stickerPackList.size), pack)
                allStickerPacksListAdapter.submitList(ArrayList(stickerPackList))
                updateEmptyState()
                supportActionBar?.title = resources.getQuantityString(
                    R.plurals.title_activity_sticker_packs_list, stickerPackList.size)
                Toast.makeText(this@StickerPackListActivity, getString(R.string.error_with_message, e.message), Toast.LENGTH_LONG).show()
            }
        }
    }


    private fun updateEmptyState() {
        if (stickerPackList.isEmpty()) {
            packRecyclerView.visibility = View.GONE
            emptyStateLayout.visibility = View.VISIBLE
            importFab.visibility = View.GONE
        } else {
            packRecyclerView.visibility = View.VISIBLE
            emptyStateLayout.visibility = View.GONE
            importFab.visibility = View.VISIBLE
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menu.add(Menu.NONE, 1001, Menu.NONE, R.string.settings)
            .setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
        menu.add(Menu.NONE, 1002, Menu.NONE, R.string.merge_packs_menu_item)
            .setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            1001 -> {
                settingsLauncher.launch(Intent(this, SettingsActivity::class.java))
                true
            }
            1002 -> {
                mergePacksLauncher.launch(Intent(this, MergeStickerPacksActivity::class.java))
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    override fun onResume() {
        super.onResume()
        refreshStickerPacks()
        if (::allStickerPacksListAdapter.isInitialized) {
            allStickerPacksListAdapter.invalidateAnimationsCache()
        }
    }

    private var globalLayoutListener: android.view.ViewTreeObserver.OnGlobalLayoutListener? = null

    private fun showStickerPackList(packList: List<StickerPack>) {
        allStickerPacksListAdapter = StickerPackListAdapter { pack -> 
            addStickerPackToWhatsApp(pack.identifier, pack.name) 
        }
        allStickerPacksListAdapter.submitList(packList)
        
        packLayoutManager = LinearLayoutManager(this)
        packLayoutManager.orientation = RecyclerView.VERTICAL
        packRecyclerView.layoutManager = packLayoutManager
        packRecyclerView.itemAnimator = null
        packRecyclerView.layoutAnimation = null
        packRecyclerView.adapter = allStickerPacksListAdapter
        packRecyclerView.setHasFixedSize(true)
        packRecyclerView.addOnScrollListener(scrollListener)
        
        globalLayoutListener = android.view.ViewTreeObserver.OnGlobalLayoutListener { recalculateColumnCount() }
        packRecyclerView.viewTreeObserver.addOnGlobalLayoutListener(globalLayoutListener)
    }

    private fun recalculateColumnCount() {
        val rvWidth = packRecyclerView.width
        if (rvWidth <= 0) return

        val density = resources.displayMetrics.density
        // Correct overhead: Card margins (12+12) + RV start (16) + RV to button (16) + button (48) + button end (16)
        val overhead = ((12 + 12 + 16 + 16 + 48 + 16) * density).roundToInt()
        val imageRowWidth = rvWidth - overhead
        if (imageRowWidth <= 0) return

        val maxNumberOfImagesInARow = STICKER_PREVIEW_DISPLAY_LIMIT
        
        // Aim for a small fixed margin between stickers
        val desiredMargin = (4 * density).roundToInt()
        val totalMargin = desiredMargin * (maxNumberOfImagesInARow - 1)
        
        // Calculate the preview size that allows exactly 5 stickers to fit
        val previewSize = (imageRowWidth - totalMargin) / maxNumberOfImagesInARow
        
        // Recalculate actual margin to distribute any rounding remainder
        val actualMargin = (imageRowWidth - maxNumberOfImagesInARow * previewSize) / (maxNumberOfImagesInARow - 1)
        
        allStickerPacksListAdapter.setImageRowSpec(maxNumberOfImagesInARow, actualMargin, previewSize)
        
        if (packRecyclerView.viewTreeObserver.isAlive) {
            packRecyclerView.viewTreeObserver.removeOnGlobalLayoutListener(globalLayoutListener)
        }
    }

    companion object {
        const val EXTRA_STICKER_PACK_LIST_DATA = "sticker_pack_list"
        private const val STICKER_PREVIEW_DISPLAY_LIMIT = 5
    }
}
